man/snmpstatus.1
.\* /***********************************************************
.\" 	Copyright 1988, 1989 by Carnegie Mellon University
.\" 
.\"                       All Rights Reserved
.\" 
.\" Permission to use, copy, modify, and distribute this software and its 
.\" documentation for any purpose and without fee is hereby granted, 
.\" provided that the above copyright notice appear in all copies and that
.\" both that copyright notice and this permission notice appear in 
.\" supporting documentation, and that the name of CMU not be
.\" used in advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.  
.\" 
.\" CMU DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
.\" ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
.\" CMU BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
.\" ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
.\" WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
.\" ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
.\" SOFTWARE.
.\" ******************************************************************/
.TH SNMPSTATUS 1 "17 September 1989"
.UC 4
.SH NAME
snmpstatus - Retrieves important information from a network entity.
.SH SYNOPSIS
snmpstatus host [community]
.SH DESCRIPTION
Snmpstatus is an SNMP application that retrieves several important
statistics from a network entity.
.PP
The
.I host
specification may be either a host name or an internet address
specified in "dot notation"
.PP
The
.I community
specifies the community name for the transaction with the remote system.
If this argument is not given, the community name will default to
.I "public".
.PP
The information returned is:
.PP
The IP address of the entity.
.br
A textual description of the entity
.I (sysDescr.0)
.br
The uptime of the entity
.I (sysUpTime.0)
.br
The sum of received packets on all interfaces
.I (ifInUCastPkts.* + ifInNUCastPkts.*)
.br
The sum of transmitted packets on all interfaces
.I (ifOutUCastPkts.* + ifOutNUCastPkts.*)
.br
The number of IP input packets
.I (ipInReceives.0)
.br
The number of IP output packets
.I (ipOutRequests.0)
.PP
For example
.PP
.I snmpstatus netdev-kbox.cc.cmu.edu public
.PP
will produce output similar to the following:
.PP
.I [128.2.56.220]=>[Kinetics FastPath2] Up: 1 day, 4:43:31
.br
.I IF recv/trans packets 262874/39867 |
.I IP recv/trans packets 31603/15805
.PP
Snmpstatus also checks the operational status of all interfaces (ifOperStatus.*),
and if it finds any that are not running, it will report in a manner similar to this:
.PP
.I 2 interfaces are down!
The
.I host
specification may be either a host name or an internet address
specified in "dot notation"
.PP
The
.I community
specifies the community name for the transaction with the remote system.  If none is
given, the community name defaults to
.I "public".
.PP
If the network entity has an error processing the request packet, an error
packet will be returned and a message will be shown, helping to pinpoint in what way
the request was malformed.  Snmpstatus will attempt to reform it's request to eliminate
the malformed variable, but this variable will then be missing from the displayed data.
.PP
.SH "SEE ALSO"
RFC 1065, RFC 1066, RFC 1067


NEWS
This file contains major changes in released revisions.  Please read
the ChangeLog file for a more detailed listing of changes/bug fixes.

*5.2.3*

    snmpusm:
      - allows to use old and/or new localized key instead of passphrase with
        option -Ck
      - new option -CE to specify usmUserEngineID (necessary to manage
        snmptrapd's usmUserTable)

*5.2.2*

 New:
   - Due to a copyright statement that didn't allow modifications, 
     snmpnetstat was completely rewritten.  The new binary is named
     snmpnetstat53, and we've provided a snmpnetstat wrapper script to
     make the command line options backwards compatible.  As a side
     effect, however, it now finally supports SNMPv3 at least.

 Ports:
   - Win32
      - Cygwin compiler fixes
      - Fix for 'select: No such file or directory' in snmpdelta

 Fixes:
   - Persistent files in directory defined by snmp.conf persistentDir were 
     not being loaded at startup
   - Perl getnext didn't honor explicit reference to MIB file
   - Better configure detection for embedded perl functions
   - Better documentation and snmpconf support for using AES encryption 
   - Many many other miscellaneous bugs

*5.2.1*
 Fixes:
   Building:
     - configure --disable-snmpv2c now works
     - fix make test tests for rfc1213
     - bug 1049607: net-snmp-config --compile-subagent broken

   library:
     - bug 1084413: Can't disable file logging
     - bug 1072406: invalid operator precedence in opendir()

   agent library:
     - disconnected AgentX subagents now reconnect with correct context
     - fix table_array row insert/delete during set processing

   agent:
     - don't override clientAddr setting for local trapsinks
     - bug 1088765: Agent fails to send traps to remote target
     - bug 1034008: memory leak using SET for table_dataset
     - patch 1052460: fix agent deadlock on exec
     - bug 1055781: get-next fails to step into interfaces group correctly
     - bug 1056760: agent ignores ifspeed, type settings in snmpd.conf
     - bug 1062986: pass and pass_persist fail and crash snmpd
     - fix snmpd.conf table token to handle augments tables

   snmptrapd:
     - bug 1085981: snmptrapd complains about logging and access control
     - bug 1040711: snmptrapd: SIGHUP duplicates traphandlers (repeatedly)

   MFD:
     - Misc updates to MFD templates
     - add auto-handling of cache update for row insert/delete

 Ports:
   - Win32
      - fixes for compiling without the Microsoft PSDK installed
      - fix Win32 getenv crash
   - Mac OS X compile error fix
   - HP-UX configure now detects and won't use unavailable function
   - Linux
      - patch 1055036: if-mib init order fix
      - patch 1057057: ipSystemStatsTable index fix, add ipv6
      - patch 1073897: fix if-mib data access 64bit counter wrap detection

 MIBs:
   - update IP-FORWARD-MIB from an ID set to become an RFC


*5.2*
 New:
    For the security paranoid (like Wes):
      - Insecure versions of SNMP may be completely disabled at compile time.
          (configure using one or both of --disable-snmpv1 --disable-snmpv2c)
      - Less secure versions of SNMPv3 authentication and encryption
          algorithms may be completely disabled at compile time.
          (configure using one or both of --disable-des --disable-md5)
          (This requires you have OpenSSL to support SHA1 and AES instead)
      - Diffie-Helman key negotiation (perfect-forward-secrecy - RFC2786)
          has been implemented in both the agent and "snmpusm".
      - Better support for direct use of localized and master keys
        (randomly generated keys have more entropy than passwords). See
        the -3m -3M -3k and -3K options, and the defAuthLocalizedKey,
        defPrivLocalizedKey, defAuthMasterKey, defPrivMasterKey
        snmp.conf tokens)

    library:
      - send and receive buffer sizes for UDP/TCP are configurable
      - .conf file token registration can now search multiple files
        (separate file names to search by a :.  IE, "snmpd:agentx")

    snmptrapd:
      - Embedded perl support for trap handlers (much faster than traphandles).
          (configure --with-embedded-perl ; see 'perldoc NetSNMP::TrapReceiver')
      - the snmptrapd usmUser table can be controlled through SNMP & AgentX
	  (the snmptrapd registers MIBs under the "snmptrapd" SNMPv3 context)

    agent:
      - community strings can be mapped to individual SNMPv3 contexts
          (see the snmpd.conf manual on "com2sec").
      - AgentX should work properly with SNMPv3 contexts now.
      - improved version of "exec" extension directive - "extend":
          supports multi-line output, configurable command input, SET
          handling, a valid relocatable MIB output structure, etc
      - more flexibility added to netsnmp_cache helper, including periodic
          refreshing of caches and cache pre-loading.
      - experimental implementation of DisMan Remote Operations MIB groups
        (remote nslookup, ping, traceroute and expressions via SNMP)
      - The proxy supports a new -Cc flag which passes the community
        name through to the other side rather than hard-coding it.
      - The config_require() and other modules can now be placed
        inside of CPP directievs (#if, /* ... */) as header files are
        now passed through CPP processing first.
      - new config_exclude() macro for excluding certain modules
      - new config_belongs_in() macro to specify if a chunk of code
        found by --with-mib-modules belongs in the agent library or
        the mib module library.
      - Uses a smaller function stack.

    commands:
      - snmpusm: A number of key-change improvements
        (EG, changing localized keys supported, diffie-helman support, ...)
      - New -Ln option to completely disable logging.

    mib2c:
      - a number of new or updated configuration templates:
        - mib2c.iterate.conf: the iterator-based table config

net-snmp/apps/snmpstatus.c
/*
 * snmpstatus.c - send snmp GET requests to a network entity.
 *
 */
/***********************************************************************
	Copyright 1988, 1989, 1991, 1992 by Carnegie Mellon University

                      All Rights Reserved

Permission to use, copy, modify, and distribute this software and its 
documentation for any purpose and without fee is hereby granted, 
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in 
supporting documentation, and that the name of CMU not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.  

CMU DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL
CMU BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR
ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS,
WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION,
ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
SOFTWARE.
******************************************************************/
#include <config.h>

#if HAVE_STDLIB_H
#include <stdlib.h>
#endif
#if HAVE_UNISTD_H
#include <unistd.h>
#endif
#if HAVE_STRING_H
#include <string.h>
#else
#include <strings.h>
#endif
#include <sys/types.h>
#if HAVE_NETINET_IN_H
#include <netinet/in.h>
#endif
#include <stdio.h>
#include <ctype.h>
#if TIME_WITH_SYS_TIME
# ifdef WIN32
#  include <sys/timeb.h>
# else
#  include <sys/time.h>
# endif
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif
#if HAVE_SYS_SELECT_H
#include <sys/select.h>
#endif
#if HAVE_WINSOCK_H
#include <winsock.h>
#endif
#if HAVE_NETDB_H
#include <netdb.h>
#endif
#if HAVE_ARPA_INET_H
#include <arpa/inet.h>
#endif

#include <getopt.h>

#include "asn1.h"
#include "snmp_api.h"
#include "snmp_impl.h"
#include "snmp_client.h"
#include "mib.h"
#include "snmp.h"
#include "system.h"
#include "tools.h"
#include "snmp_parse_args.h"
#include "default_store.h"
#include "snmp_transport.h"

oid	objid_sysDescr[] = {1, 3, 6, 1, 2, 1, 1, 1, 0};
size_t	length_sysDescr = sizeof(objid_sysDescr)/sizeof(oid);
oid	objid_sysUpTime[] = {1, 3, 6, 1, 2, 1, 1, 3, 0};
size_t	length_sysUpTime = sizeof(objid_sysUpTime)/sizeof(oid);
oid	objid_ifOperStatus[] = {1, 3, 6, 1, 2, 1, 2, 2, 1, 8};
size_t	length_ifOperStatus = sizeof(objid_ifOperStatus)/sizeof(oid);
oid	objid_ifInUCastPkts[] = {1, 3, 6, 1, 2, 1, 2, 2, 1, 11};
size_t	length_ifInUCastPkts = sizeof(objid_ifInUCastPkts)/sizeof(oid);
oid	objid_ifInNUCastPkts[] = {1, 3, 6, 1, 2, 1, 2, 2, 1, 12};
size_t	length_ifInNUCastPkts = sizeof(objid_ifInNUCastPkts)/sizeof(oid);
oid	objid_ifOutUCastPkts[] = {1, 3, 6, 1, 2, 1, 2, 2, 1, 17};
size_t	length_ifOutUCastPkts = sizeof(objid_ifOutUCastPkts)/sizeof(oid);
oid	objid_ifOutNUCastPkts[] = {1, 3, 6, 1, 2, 1, 2, 2, 1, 18};
size_t	length_ifOutNUCastPkts = sizeof(objid_ifOutNUCastPkts)/sizeof(oid);
oid	objid_ipInReceives[] = {1, 3, 6, 1, 2, 1, 4, 3, 0};
size_t	length_ipInReceives = sizeof(objid_ipInReceives)/sizeof(oid);
oid	objid_ipOutRequests[] = {1, 3, 6, 1, 2, 1, 4, 10, 0};
size_t	length_ipOutRequests = sizeof(objid_ipOutRequests)/sizeof(oid);

#define DS_APP_DONT_FIX_PDUS 0

static void optProc(int argc, char *const *argv, int opt)
{
    switch (opt) {
        case 'C':
            while (*optarg) {
                switch (*optarg++) {
                    case 'f':
                        ds_toggle_boolean(DS_APPLICATION_ID, DS_APP_DONT_FIX_PDUS);
                        break;
		    default:
		        fprintf(stderr, "Unknown flag passed to -C: %c\n", optarg[-1]);
			exit(1);
                }
            }
            break;
    }
}

void usage(void)
{
  fprintf(stderr,"Usage: snmpstatus [-Cf]");
  snmp_parse_args_usage(stderr);
  fprintf(stderr,"\n\n");
  snmp_parse_args_descriptions(stderr);
  fprintf(stderr, "snmpstatus specific options\n");
  fprintf(stderr, "  -Cf\t\tDon't fix errors and retry the request.\n");
}


int main(int argc, char *argv[])
{
    struct snmp_session session, *ss;
    struct snmp_pdu *pdu, *response;
    struct variable_list *vars;
    snmp_transport *transport = NULL;
    char *sysdescr = NULL;
    u_long uptime = 0;
    int status;
    int ipin = 0, ipout = 0, ipackets = 0, opackets = 0;
    int good_var;
    int down_interfaces = 0;
    char buf[40];
    int interfaces;
    int count;
    int exitval = 0;

    /* get the common command line arguments */
    switch (snmp_parse_args(argc, argv, &session, "C:", &optProc)) {
    case -2:
    	exit(0);
    case -1:
        usage();
        exit(1);
    default:
        break;
    }

    SOCK_STARTUP;

    /* open an SNMP session */
    ss = snmp_open(&session);
    if (ss == NULL){
      /* diagnose snmp_open errors with the input struct snmp_session pointer */
      snmp_sess_perror("snmpstatus", &session);
      SOCK_CLEANUP;
      exit(1);
    }

    /* create PDU for GET request and add object names to request */
    pdu = snmp_pdu_create(SNMP_MSG_GET);
    snmp_add_null_var(pdu, objid_sysDescr, length_sysDescr);
    snmp_add_null_var(pdu, objid_sysUpTime, length_sysUpTime);
    snmp_add_null_var(pdu, objid_ipInReceives, length_ipInReceives);
    snmp_add_null_var(pdu, objid_ipOutRequests, length_ipOutRequests);

    /* do the request */
retry:
    status = snmp_synch_response(ss, pdu, &response);
    if (status == STAT_SUCCESS){
      if (response->errstat == SNMP_ERR_NOERROR){
        for(vars = response->variables; vars; vars = vars->next_variable){
          if (vars->name_length == length_sysDescr &&
              !memcmp(objid_sysDescr, vars->name, sizeof(objid_sysDescr))){
            sysdescr = (char*)malloc(vars->val_len+1);
            memcpy(sysdescr, vars->val.string, vars->val_len);
            sysdescr[vars->val_len] = '\0';
          }
          if (vars->name_length == length_sysUpTime &&
              !memcmp(objid_sysUpTime, vars->name, sizeof(objid_sysUpTime))){
            uptime = *vars->val.integer;
          }
          if (vars->name_length == length_ipInReceives &&
              !memcmp(objid_ipInReceives, vars->name, sizeof(objid_ipInReceives))){
            ipin = *vars->val.integer;
          }
          if (vars->name_length == length_ipOutRequests &&
              !memcmp(objid_ipOutRequests, vars->name, sizeof(objid_ipOutRequests))){
            ipout = *vars->val.integer;
          }
        }
      } else {
        fprintf(stderr, "Error in packet.\nReason: %s\n",
                snmp_errstring(response->errstat));
        if (response->errindex != 0){
          fprintf(stderr, "Failed object: ");
          for(count = 1, vars = response->variables;
                vars && count != response->errindex;
                vars = vars->next_variable, count++)
            ;
          if (vars)
            fprint_objid(stderr, vars->name, vars->name_length);
          fprintf(stderr, "\n");
        }

        /* retry if the errored variable was successfully removed */
        if (!ds_get_boolean(DS_APPLICATION_ID, DS_APP_DONT_FIX_PDUS)) {
            pdu = snmp_fix_pdu(response, SNMP_MSG_GET);
            snmp_free_pdu(response);
            response = NULL;
            if (pdu != NULL)
                goto retry;
        }
      }
    } else if (status == STAT_TIMEOUT){
      fprintf(stderr,"Timeout: No Response from %s\n", session.peername);
      SOCK_CLEANUP;
      exit(1);
    } else {    /* status == STAT_ERROR */
      snmp_sess_perror("snmpstatus", ss);
      SOCK_CLEANUP;
      exit(2);
    }

    transport = snmp_sess_transport(snmp_sess_pointer(ss));
    if (transport != NULL && transport->f_fmtaddr != NULL) {
      char *s = transport->f_fmtaddr(transport,
				     response->transport_data,
				     response->transport_data_length);
      printf("[%s]=>[%s] Up: %s\n", s, sysdescr, uptime_string(uptime, buf));
    } else {
      printf("[<UNKNOWN>]=>[%s] Up: %s\n", sysdescr,
	     uptime_string(uptime, buf));
    }

    if (response)
      snmp_free_pdu(response);

    /* create PDU for GET request and add object names to request */
    pdu = snmp_pdu_create(SNMP_MSG_GETNEXT);
    snmp_add_null_var(pdu, objid_ifOperStatus, length_ifOperStatus);
    snmp_add_null_var(pdu, objid_ifInUCastPkts, length_ifInUCastPkts);
    snmp_add_null_var(pdu, objid_ifInNUCastPkts, length_ifInNUCastPkts);
    snmp_add_null_var(pdu, objid_ifOutUCastPkts, length_ifOutUCastPkts);
    snmp_add_null_var(pdu, objid_ifOutNUCastPkts, length_ifOutNUCastPkts);

/*?? note: this code is not quite complete */
    good_var = 5;
    interfaces = 0;
    while(good_var == 5){
      good_var = 0;
      status = snmp_synch_response(ss, pdu, &response);
      if (status == STAT_SUCCESS){
        if (response->errstat == SNMP_ERR_NOERROR){
          pdu = snmp_pdu_create(SNMP_MSG_GETNEXT);
          for(vars = response->variables; vars; vars = vars->next_variable){
            if (vars->name_length >= length_ifOperStatus &&
                !memcmp(objid_ifOperStatus, vars->name,
                        sizeof(objid_ifOperStatus))){
              if (*vars->val.integer != MIB_IFSTATUS_UP)
                down_interfaces++;
              snmp_add_null_var(pdu, vars->name, vars->name_length);
              good_var++;
            } else if (vars->name_length >= length_ifInUCastPkts &&
                !memcmp(objid_ifInUCastPkts, vars->name,
                        sizeof(objid_ifInUCastPkts))){
              ipackets += *vars->val.integer;
              snmp_add_null_var(pdu, vars->name, vars->name_length);
              good_var++;
            } else if (vars->name_length >= length_ifInNUCastPkts &&
                !memcmp(objid_ifInNUCastPkts, vars->name,
                        sizeof(objid_ifInNUCastPkts))){
              ipackets += *vars->val.integer;
              snmp_add_null_var(pdu, vars->name, vars->name_length);
              good_var++;
            } else if (vars->name_length >= length_ifOutUCastPkts &&
                !memcmp(objid_ifOutUCastPkts, vars->name,
                        sizeof(objid_ifOutUCastPkts))){
              opackets += *vars->val.integer;
              snmp_add_null_var(pdu, vars->name, vars->name_length);
              good_var++;
            } else if (vars->name_length >= length_ifOutNUCastPkts &&
                !memcmp(objid_ifOutNUCastPkts, vars->name,
                        sizeof(objid_ifOutNUCastPkts))){
              opackets += *vars->val.integer;
              snmp_add_null_var(pdu, vars->name, vars->name_length);
              good_var++;
            }
          }
          if (good_var == 5)
            interfaces++;
        } else {
          fprintf(stderr, "Error in packet.\nReason: %s\n",
                  snmp_errstring(response->errstat));
          if (response->errindex != 0){
            fprintf(stderr, "Failed object: ");
            for(count = 1, vars = response->variables;
                  vars && count != response->errindex;
                  vars = vars->next_variable, count++)
              ;
            if (vars)
              fprint_objid(stderr, vars->name, vars->name_length);
            fprintf(stderr, "\n");
          }
	  exitval = 2;
        }
      } else if (status == STAT_TIMEOUT){
        fprintf(stderr,"Timeout: No Response from %s\n", session.peername);
	exitval = 1;
      } else {    /* status == STAT_ERROR */
        snmp_sess_perror("snmpstatus", ss);
	exitval = 1;
      }

      if (response)
        snmp_free_pdu(response);
    }
    printf("Interfaces: %d, Recv/Trans packets: %d/%d | IP: %d/%d\n",
            interfaces, ipackets, opackets, ipin, ipout);
    if (down_interfaces > 0){
      printf("%d interface%s down!\n",
            down_interfaces, down_interfaces > 1 ? "s are": " is" );
    }

    snmp_close(ss);
    SOCK_CLEANUP;
    return exitval;
}
